README.md
# Net::FTPSSL::Robust

  * My extended documentation: <http://perl.overmeer.net/CPAN/>
  * Development via GitHub: <https://github.com/markov2/perl5-Net-FTPSSL-Robust>
  * Download from CPAN: <ftp://ftp.cpan.org/pub/CPAN/authors/id/M/MA/MARKOV/>
  * Indexed from CPAN: <http://search.cpan.org/~markov/Net-FTPSSL-Robust/>
    and <https://metacpan.org/release/Net-FTPSSL-Robust>

Downloading large quantities of information require recovery: the
chance that your task fails runs larger when it runs longer.  This
module implements retries on top of Net::FTPSSL.

Also see Net::FTP::Robust.

## Development &rarr; Release

Important to know, is that I use an extension on POD to write the manuals.
The "raw" unprocessed version is visible on GitHub.  It will run without
problems, but does not contain manual-pages.

Releases to CPAN are different: "raw" documentation gets removed from
the code and translated into real POD and clean HTML.  This reformatting
is implemented with the OODoc distribution (A name I chose before OpenOffice
existed, sorry for the confusion)

Clone from github for the "raw" version.  For instance, when you want
to contribute a new feature.

On github, you can find the processed version for each release.  But the
better source is CPAN; to get it installed simply run:

```sh
   cpan -i Net::FTPSSL::Robust
```

## Contributing

When you want to contribute to this module, you do not need to provide
a perfect patch... actually: it is nearly impossible to create a patch
which I will merge without modification.  Usually, I need to adapt the
style of code and documentation to my own strict rules.

When you submit an extension, please contribute a set with

1. code

2. code documentation

3. regression tests in t/

**Please note:**
When you contribute in any way, you agree to transfer the copyrights to
Mark Overmeer (you will get the honors in the code and/or ChangeLog).
You also automatically agree that your contribution is released under
the same license as this project: licensed as perl itself.

## Copyright and License

This project is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.
See <http://dev.perl.org/licenses/>



ChangeLog

===== version history for Net::FTPSSL::Robust

All changes made by Mark Overmeer, unless explicitly stated
differently.

version 0.07:

	Fixes:
	- fix metadata [Mohammad S Anwar]

version 0.06: Mon Jan 22 21:38:55 CET 2018

	Improvements:
	- convert to GIT
	- move to GitHUB

version 0.05: Mon Aug 22 10:27:50 CEST 2011
	- Net::FTP::SSL >= 0.18 supports restart.
	  rt.cpan.org [Curtis Leach]
	- move t/99pod.t to xt/ so we can remove dependency on Test::POD

version 0.04: Mon Nov 16 21:42:35 CET 2009
	- another fix :(  $ftp->nlst incompatible with Net::FTP

version 0.03: Mon Nov 16 21:02:19 CET 2009
	- change bin/ftpssl-get meaning for 'p' <-> 'P' to be
	  compatible with bin/ftp-get.

version 0.02: Fri Nov 13 23:42:02 CET 2009
	- ftpssl-get change option flag abbreviation of "protection"
	  from 'p' into 'd', because "port" used 'p' as well.
	- And "password" abbreviation became capital 'P'.
	- do not attempt to restart

version 0.01: Fri Nov 13 17:02:09 CET 2009
	- initial version


lib/Net/FTPSSL/Robust.pm
# This code is part of distribution Net-FTPSSL-Robust.  Meta-POD processed
# with OODoc into POD and HTML manual-pages.  See README.md
# Copyright Mark Overmeer.  Licensed under the same terms as Perl itself.

package Net::FTPSSL::Robust;
use base 'Net::FTP::Robust', 'Exporter';

use warnings;
use strict;

our @EXPORT = qw/SECURITY_TLS SECURITY_SSL/;

use Log::Report 'net-ftpssl-robust', syntax => 'SHORT';

use Net::FTPSSL;
# Gladly, ::FTPSSL has some level of interface compatibility with Net::FTP

use constant
 { SECURITY_TLS => 'TLS'
 , SECURITY_SSL => 'SSL'
 };

=chapter NAME

Net::FTPSSL::Robust - download files over FTP

=chapter SYNOPSIS

 use Net::FTPSSL;          # will publish some constants
 use Net::FTPSSL::Robust;  # more constants

 my $ftp = Net::FTPSSL::Robust->new
  ( Host     => $host
  , Security => SECURITY_TLS
  );

 # Further, see Net::FTP::Robust

=chapter DESCRIPTION

This module is specialized in (on the moment only) downloading large
data-sets (gigabytes) via FTP over an encrypted connection. It uses
M<Net::FTPSSL> for the encryption and M<Net::FTP::Robust> for the
logic.

=chapter METHODS

=section Constructors

=c_method new OPTIONS

Use to connect to one ftp-server.
All OPTIONS are passed to M<Net::FTPSSL> method C<new()>. These options
all start with capitals.

=option  useSSL BOOLEAN
=default useSSL <false>
When true SSL is used, otherwise TLS. This is the only M<Net::FTPSSL>
option which starts with a lower-case. Usable, but deprecated: please
use 

=option  Security SECURITY_TLS|SECURITY_SSL
=default Security SECURITY_TLS
Replaces the C<useSSL> option for M<Net::FTPSSL>.  The constants are
exported.
=cut

sub init($)
{   my ($self, $args) = @_;

    if(my $sec = delete $args->{Security})
    {   $args->{useSSL} =
            $sec eq SECURITY_TLS ? 0
          : $sec eq SECURITY_SSL ? 1
          : error "unknown security protocol {proto}", proto => $sec;
    }

    $self->SUPER::init($args);
    $self;
}

=section Download

=cut

sub _connect($)
{   my ($self, $opts) = @_;
    my $host = $opts->{Host}
        or error "no host provided to connect to";

    my $ftp = Net::FTPSSL->new($host, %$opts);
    my $err = defined $ftp ? undef : $Net::FTPSSL::ERRSTR;
    ($ftp, $err);
}

sub _modif_time($$)
{   my ($self, $ftp, $fn) = @_;
    $ftp->_mdtm($fn);
}

sub _ls($) { $_[1]->nlst }

sub _can_restart($$$) { 1 }

=chapter DETAILS

=section Comparison

M<Net::FTPSSL> implements the FTP protocol over encrypted connections.
C<Net::FTPSSL::Robust> adds retries and logging, to retrieve
data which takes hours to download, sometimes over quite instable
connections.  It uses M<Log::Report> which can connect to various logging
frameworks for its messages.

=section Limitations

See L<Net::FTP::Robust/Limitations>.

=cut

1;


Makefile.PL
use ExtUtils::MakeMaker;

use 5.008;

WriteMakefile
 ( NAME	 	=> 'Net::FTPSSL::Robust'
 , VERSION	=> '0.07'
 , PREREQ_PM    =>
   { Log::Report      => 0.25
   , Net::FTP::Robust => 0.06
   , Net::FTPSSL      => 0.18
   }
 , EXE_FILES    => [ 'bin/ftpssl-get' ]
 , AUTHOR       => 'Mark Overmeer <markov@cpan.org>'
 , ABSTRACT     => 'download directories over FTP with SSL or TLS encryption'
 , LICENSE      => 'perl_5'

 , META_MERGE =>
      { 'meta-spec' => { version => 2 }
      , resources  =>
          { repository =>
              { type => 'git'
              , url  => 'https://github.com/markov2/perl5-Net-FTPSSL-Robust.git'
              , web  => 'https://github.com/markov2/perl5-Net-FTPSSL-Robust'
              }
          , homepage => 'http://perl.overmeer.net/CPAN/'
          , license  => [ 'http://dev.perl.org/licenses/' ]
          }
      }

 );

### used by oodist during production of distribution
sub MY::postamble { <<'__POSTAMBLE' }

# for DIST
RAWDIR          = ../public_html/net-ftpssl-robust/raw
DISTDIR         = ../public_html/net-ftpssl-robust/source
LICENSE         = perl

# for POD
FIRST_YEAR      = 2009
EMAIL           = markov@cpan.org
WEBSITE         = http://perl.overmeer.net/CPAN/
__POSTAMBLE
